drivers/scheduler/anthos/anthos.go

```diff
@@ -529,6 +544,122 @@ func (anth *anthos) unsetUserNameAndKey() error {
 	return nil
 }
 
+// checkUserClusterNodesUpgradeTime measure the time taken by each node and report error
+func (anth *anthos) checkUserClusterNodesUpgradeTime() error {
+	log.Info("Validating user cluster nodes upgrade time")
+	userCluster, err := anth.getUserClusterName()
+	if err != nil {
+		return err
+	}
+	initNodeUpgradeTime, err := anth.getStartTimeForNodePoolUpgrade(userCluster)
+	if err != nil {
+		return err
+	}
+	log.Debugf("User cluster node pool upgrade started at: [%v]", initNodeUpgradeTime.Format(time.UnixDate))
+	poolMap, err := getNodesSortByAge()
+	if err != nil {
+		return err
+	}
+	for _, sortedNodes := range poolMap {
+		startTime := initNodeUpgradeTime
+		for _, node := range sortedNodes {
+			diff := node.CreationTimestamp.Sub(startTime)
+			if diff > errorTimeDuration {
+				return fmt.Errorf("[%s] node upgrade took: [%v] minutes which is longer than the expected timeout value: [%v]",
+					node.Name, diff, errorTimeDuration)
+			}
+			log.Infof("[%s] node took: [%v] time to upgrade the node", node.Name, diff)
+			startTime = node.CreationTimestamp.Time
+		}
+	}
+	return nil
+}
+
+// getUserClusterName return Anthos user cluster name
+func (anth *anthos) getUserClusterName() (string, error) {
+	log.Info("Retrieving user cluster name")
+	var userCluster string
+	// Listing user cluster to get user cluster name
+	cmd := fmt.Sprintf("%s --kubeconfig %s clusters |grep -v NAME", listUserClustersCmd, adminKubeconfPath)
+	out, err := anth.execOnAdminWSNode(cmd)
+	if err != nil {
+		return "", fmt.Errorf("listing user clusters is failing: [%s]. Err: (%v)", out, err)
+	}
+	userClusters := strings.Split(out, "\n")
+	for _, cluster := range userClusters {
+		clusterInfo := strings.Fields(cluster)
+		userCluster = clusterInfo[1]
+		break
+	}
+	if userCluster == "" {
+		return "", fmt.Errorf("failed to find user cluster name")
+	}
+	log.Infof("Successfully retrieved user cluster name: [%s]", userCluster)
+	return userCluster, nil
+}
+
+// getStartTimeForNodePoolUpgrade return start time when node pool upgrade started
+func (anth *anthos) getStartTimeForNodePoolUpgrade(userClusterName string) (time.Time, error) {
+	log.Info("Getting start time for node pool upgrade")
+	var timeMatchReg = regexp.MustCompile(`\d+-\d+-\d+T\d+:\d+:\d+Z`)
+	var nodeUpgradeStartedReg = regexp.MustCompile(`gke-on-prem-last-upgrade-start-time: .+`)
+	const layout = "2006-01-02T15:04:05Z"
```

Comment:
Does this have to be a const?

Reply:
Yes const is not needed, Incorporated !!